SparseGrids/tsgCacheLagrange.hpp

```diff
@@ -138,101 +138,58 @@ class CacheLagrangeDerivative : public CacheLagrange<T> {
      *   and the actual nodes with the pre-computed Lagrange coefficients
      * - \b holds the coordinates of the canonical point to cache
      */
-    CacheLagrangeDerivative(int num_dimensions, const std::vector<int> &max_levels, const OneDimensionalWrapper &rule, const double x[]) :
-            CacheLagrange<T>(num_dimensions, max_levels, rule, x) {
-        cacheDerivatives.resize(num_dimensions);
+    CacheLagrangeDerivative(int num_dimensions, const std::vector<int> &max_levels, const OneDimensionalWrapper &rule, const double x[]) {
+        offsets = rule.getPointsCount();
+        cache.resize(num_dimensions);
         for(int dim=0; dim<num_dimensions; dim++){
-            cacheDerivatives[dim].resize(offsets[max_levels[dim] + 1]);
+            cache[dim].resize(offsets[max_levels[dim] + 1]);
             for(int level=0; level <= max_levels[dim]; level++)
-                cacheDerivativeLevel(level, x[dim], rule, &(cache[dim][offsets[level]]), &(cacheDerivatives[dim][offsets[level]]));
+                cacheDerivativeLevel(level, x[dim], rule, &(cache[dim][offsets[level]]));
         }
     }
     //! \brief Destructor, clear all used data.
     ~CacheLagrangeDerivative() = default;
 
     //! \brief Computes the derivatives of all Lagrange polynomials for the given level at the given x
-    static void cacheDerivativeLevel(int level, double x, const OneDimensionalWrapper &rule, T *vals, T *cc){
-        // Initialize.
+    static void cacheDerivativeLevel(int level, double x, const OneDimensionalWrapper &rule, T *cc){
+        /*
+         * The usual j-th Lagrange derivative is of the form Lj(x) = cj * fj(x) * gj(x) where
+         *
+         *   fj(x) = (x - nodes[0])...(x - nodes[j-1]),
+         *   gj(x) = (x - nodes[j+1])...(x - nodes[num_points-1]),
+         *
+         * and cj is the j-th Lagrange coefficient. The product rule gives Lj'(x) = cj * [fj'(x) * gj(x) + fj(x) * gj'(x)].
+         * Hence, we do a forward pass for [fj(x), fj'(x)] and a backward pass for [gj(x), gj'(x)].
+         */
         const double *nodes = rule.getNodes(level);
         const double *coeff = rule.getCoefficients(level);
         int num_points = rule.getNumPoints(level);
-
-        // Check to see if x is one of the cached roots.
-        int match_idx = -1;
-        for (int j=0; j<num_points; j++) {
-            if (std::fabs(x - nodes[j]) <= Maths::num_tol) {
-                match_idx = j;
-                break;
-            }
+        // cc first stores fj'(x), aux_f stores fj(x), and aux_g stores gj(x).
+        T aux_f[num_points], aux_g[num_points];
```

Comment:
You cannot do this. There is a GCC extension that allows the declaration of an array with length that depends on a runtime variable, but it is not allowed in the C++ standard. Use `std::vector<T>`

Reply:
Changed.